### Pokemon and Stats EDA Project

source: https://www.kaggle.com/datasets/abcsds/pokemon/data

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re

In [40]:
#import csv
df = pd.read_csv("Pokemon.csv")
df.head()


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

## Data Cleaning


In [41]:
df.describe()

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

In [42]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


In [43]:
print(df.columns)

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')


In [44]:
duplicated = df.duplicated(keep = False)
df[duplicated]

Empty DataFrame
Columns: [#, Name, Type 1, Type 2, Total, HP, Attack, Defense, Sp. Atk, Sp. Def, Speed, Generation, Legendary]
Index: []

There are no duplicates


In [45]:
df.isnull().sum()

#               0
Name            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

Shows that there is a total of 386 null values in 'Type 2' column. Next we shall explore "Type 2" column to see if they really are null

In [46]:
df[df["Type 2"].isnull()].head(5)


#        Name Type 1 Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
4   4  Charmander   Fire    NaN    309  39      52       43       60       50   
5   5  Charmeleon   Fire    NaN    405  58      64       58       80       65   
9   7    Squirtle  Water    NaN    314  44      48       65       50       64   
10  8   Wartortle  Water    NaN    405  59      63       80       65       80   
11  9   Blastoise  Water    NaN    530  79      83      100       85      105   

    Speed  Generation  Legendary  
4      65           1      False  
5      80           1      False  
9      43           1      False  
10     58           1      False  
11     78           1      False

From my knowledge of pokemon we can determine that the null values mean that they do not have a Type 2 element. We shall replace these values with a 'Empty'

In [47]:
df['Type 2'].fillna('Empty', inplace = True)
df.head()



C:\Users\Ricky\AppData\Local\Temp\ipykernel_12332\3208040192.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Type 2'].fillna('Empty', inplace = True)


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire   Empty    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Confirm that the fillna worked

In [48]:
df.isnull().sum()

#             0
Name          0
Type 1        0
Type 2        0
Total         0
HP            0
Attack        0
Defense       0
Sp. Atk       0
Sp. Def       0
Speed         0
Generation    0
Legendary     0
dtype: int64

In [49]:
df[df['Name'] == 'Charmander']

#        Name Type 1 Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
4  4  Charmander   Fire  Empty    309  39      52       43       60       50   

   Speed  Generation  Legendary  
4     65           1      False

We can confirm that the fill na has worked. Next I noticed that there was some problems with the name, specifically mega pokemon, where the name was repeated twice for example

In [50]:
mega = df['Name'].str.contains('Mega')
df[mega].head(3)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
3  3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8  6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
3      122      120     80           1      False  
7      130       85    100           1      False  
8      159      115    100           1      False

We can see that the name of the pokemon is repeat twice. Hence we need to clean up the name by removing the first duplication

In [51]:
#(\w+) captures the first word e.g Venusaur
#(Mega \w+) captures the 'Mega ' and pokemon name e.g Mega Venusuar
# reaplces it with Mega and first capture group (\w+)
dupes_name = df['Name'].str.replace(r'(\w+)(Mega \w+)', r'Mega \1', regex = True)
dupes_name

0                Bulbasaur
1                  Ivysaur
2                 Venusaur
3            Mega Venusaur
4               Charmander
              ...         
795                Diancie
796           Mega Diancie
797    HoopaHoopa Confined
798     HoopaHoopa Unbound
799              Volcanion
Name: Name, Length: 800, dtype: object

In [52]:
df['Name'] = dupes_name

In [54]:
mega = df['Name'].str.contains('Mega')
df[mega].head(3)

#              Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
3  3     Mega Venusaur  Grass  Poison    625  80     100      123      122   
7  6  Mega Charizard X   Fire  Dragon    634  78     130      111      130   
8  6  Mega Charizard Y   Fire  Flying    634  78     104       78      159   

   Sp. Def  Speed  Generation  Legendary  
3      120     80           1      False  
7       85    100           1      False  
8      115    100           1      False

As we can see, we have removed the duplicated pokemon name in mega pokemon. However, I just saw that some pokemon still have their names duplicated such as hoopa.

In [65]:
hoopa = df['Name'].str.contains('Hoopa')
df[hoopa]

#                 Name   Type 1 Type 2  Total  HP  Attack  Defense  \
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
797      150      130     70           6       True  
798      170      130     80           6       True

We shall see if there are anymore duplicate names

In [67]:
dupes = df[df['Name'].str.contains(r'(\b\w+)\1.')]
dupes

C:\Users\Ricky\AppData\Local\Temp\ipykernel_12332\996668843.py:1: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  dupes = df[df['Name'].str.contains(r'(\b\w+)\1.')]


#                 Name   Type 1 Type 2  Total  HP  Attack  Defense  \
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
797      150      130     70           6       True  
798      170      130     80           6       True

There is 3 observation where the name is repeated. We shall remove one of the repeated names.

In [70]:
dupes_name = df['Name'].str.replace(r'HoopaHoopa (.+)', r'Hoopa \1', regex = True)
dupes_name

0           Bulbasaur
1             Ivysaur
2            Venusaur
3       Mega Venusaur
4          Charmander
            ...      
795           Diancie
796      Mega Diancie
797    Hoopa Confined
798     Hoopa Unbound
799         Volcanion
Name: Name, Length: 800, dtype: object

In [73]:
df['Name'] = dupes_name
df[df['Name'].str.contains(r'Hoopa.+')]

#            Name   Type 1 Type 2  Total  HP  Attack  Defense  Sp. Atk  \
797  720  Hoopa Confined  Psychic  Ghost    600  80     110       60      150   
798  720   Hoopa Unbound  Psychic   Dark    680  80     160       60      170   

     Sp. Def  Speed  Generation  Legendary  
797      130     70           6       True  
798      130     80           6       True

As we can see we have fixed the hoopa name duplicated

### EDA

In [1]:
#see value counts for each Type 1 
df['Type 1'].value_counts()

NameError: name 'df' is not defined

In [ ]:
total_count_by_type = df['Type 1'].value_counts() # returns a pandas series
plt.figure(figsize = (8,6))
plt.bar(total_count_by_type.index, total_count_by_type.values)
plt.xlabel('Type')
plt.ylabel('Total')
plt.xticks(rotation = 45)
plt.show()

In [84]:
no_legendaries = df[df['Legendary'] == True].shape[0]
print('There are ' , no_legendaries , ' legendaries or ' , str((no_legendaries/800)*100) +
      '% are legendaries')

There are  65  legendaries or  8.125% are legendaries
